File: HW8/171044039_CSE241_HW08/Iterator.java
/** 
 * Represents the Iterator class.
 * @see Collection  
 * @param E object-type to hold in set.
 * @author Ahmed Semih Özmekik
 * @version 1.0
 * @since 2019-01-14
*/


public class Iterator<E>
{
	private E[][] buffer; // underlying buffer.
	private int idx; // location of iterator in buffer.
	private E last_element; // last element returned by this iterator.

	public Iterator(E[][] arr)
	{
		buffer = arr;
		idx = 0;
	}

	/** 
	 * Returns true if the iteration has more elements.
	 * @return true if iteration has more elements.
	*/
	public boolean hasNext()
	{
		return idx<buffer[0].length;
	}

	/**
	 * Returns the next element in the iteration and advances the
 	 * iterator.
 	 * @return E, the next element in the iteration.
 	*/
	public E next()
	{
		last_element = buffer[0][idx++];
		return last_element;
	}
	
	/**
	 * Removes from the underlying collection the last element returned
	 * by this iterator.
	*/
	public void remove()
	{
		@SuppressWarnings("unchecked")
		E[] newArr = (E[]) new Object[buffer[0].length -1];

		for(int i=0, k=0;i<newArr.length;++i)
		{

			if(last_element == buffer[0][i])
				continue;
			
			newArr[k++] = buffer[0][i];
		}

		//change the real reference to new array.
		buffer[0] = newArr; 
	}


}

File: HW6/171044039_CSE241_HW06/Iterator.cpp
#ifndef OZMEKIK_COLLECTION_ITERATOR_H
#define OZMEKIK_COLLECTION_ITERATOR_H

#include <vector>
#include <exception>

namespace ozmekik_collection
{
	class bad_iter; // forward declaration of exception.

	template <typename E, typename container=std::vector<E> >
	class Iterator
	{	
		public:
			Iterator(container*);
			Iterator();

			const E& operator*()const; // dereference as an lvalue.
			E& operator*(); // dereference as an rvalue.

			/* Returns the next element in the iteration and advances the
			 * iterator. */
			E& next();

			/* Removes from the underlying collection the last element returned
			 * by this iterator. */
			void remove()throw(bad_iter);

			/* Returns true if the iteration has more elements. */
			bool hasNext()const;

			bool isNull()const { return (it==cp->end());}

		private:
			// Pointer to buffer. 
			container *cp; 
			// underlying std iterator to handle stuff..
			typename container::iterator it;

			// keeps track of last returned Iterator
			typename container::iterator last_returned;


	};

	template <typename E, typename container > 
	Iterator<E,container>::Iterator(): cp(nullptr),it(), last_returned()
	{/* intentionally left blank*/}

	template <typename E, typename container >
	Iterator<E,container>::Iterator(container* c)
	:cp(c), 
	 it(c->begin()), 
	 last_returned(c->end()) // Null value for iterator
	{/* intentionally left blank*/}

	template <typename E, typename container >
	E& Iterator<E,container>::next()
	{
		++it;
		last_returned = it;
		return *(*this);	
	}

	template<typename E, typename container > 
	const E& Iterator<E,container>::operator*()const
	{
		last_returned = it;
		return *it;  
	}

	template <typename E, typename container >
	E& Iterator<E,container>::operator*()
	{
		last_returned = it;
		return *it;
	} 

	template <typename E, typename container >
	bool Iterator<E,container>::hasNext()const
	{
		return (std::next(it,1)!=cp->end());
	}

	template <typename E, typename container >
	void Iterator<E,container>::remove()throw(bad_iter)
	{
		// nothing returned in session.
		if(last_returned==cp->end()) throw bad_iter();

		cp->erase(last_returned);
	}


	/* ozmekik_collection::for_each implementation.
	 * Similar to stl::for_each.(probably not implementational wise)
	 * I needed a function to iterate over my collection for 
	 * the lots of the cases. So for the sake of reuse, I added this one.
	 */ 
	template<typename E, typename container, typename F> 
	void for_each(Iterator<E,container> begin, F&& fun)
	{

	
		do // does the given unary functor to elements 
		{  //from starting iterator 'til the end.
			fun(*begin);
			begin.next();
		}while(begin.hasNext());

		fun(*begin);
		
	}


	class bad_iter : public std::exception
	{
		public:
			virtual const char* what() const noexcept override
			{
				return "Can't Remove! There isn't any last returned iterator.";
			}
	};
	


}

#endif // OZMEKIK_COLLECTION_ITERATOR_H